### PR TITLE
PPTP-363: initial sole trader integration

### PR DESCRIPTION
**Pull request by tombart**

### CODE REVIEW

File: app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ContactDetailsConfirmAddressController.scala [hunk 1]
```diff
@@ -50,26 +52,51 @@ class ContactDetailsConfirmAddressController @Inject() (
 )(implicit ec: ExecutionContext)
     extends FrontendController(mcc) with Cacheable with I18nSupport {
 
+  //TODO - temporary while we're working out where to get it from
```
Comment (PauloPortugal) [inline app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ContactDetailsConfirmAddressController.scala:55]: :+1:

File: app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ContactDetailsConfirmAddressController.scala [hunk 2]
```diff
@@ -50,26 +52,51 @@ class ContactDetailsConfirmAddressController @Inject() (
...
         case _ => Future(Redirect(routes.RegistrationController.displayPage()))
       }
     }
 
+  private def executeAddressUpdate(
+    journeyId: String,
+    getAddressFunction: (String, JourneyRequest[AnyContent], HeaderCarrier) => Future[Address]
+  )(implicit request: JourneyRequest[AnyContent]) =
+    for {
+      updatedAddress <- getAddressFunction(journeyId, request, hc)
```
Comment (PauloPortugal) [inline app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ContactDetailsConfirmAddressController.scala:91]: Potentially pass the `request` and `hc` as implicits to this method.

File: app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ContactDetailsConfirmAddressController.scala [hunk 3]
```diff
@@ -50,26 +52,51 @@ class ContactDetailsConfirmAddressController @Inject() (
...
+          }
+
         case _ => Future(Redirect(routes.RegistrationController.displayPage()))
       }
     }
 
+  private def executeAddressUpdate(
+    journeyId: String,
+    getAddressFunction: (String, JourneyRequest[AnyContent], HeaderCarrier) => Future[Address]
+  )(implicit request: JourneyRequest[AnyContent]) =
```
Comment (PauloPortugal) [inline app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ContactDetailsConfirmAddressController.scala:89]: We could also pass the `hc` as implicit

File: app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ContactDetailsConfirmAddressController.scala [hunk 4]
```diff
@@ -50,26 +52,51 @@ class ContactDetailsConfirmAddressController @Inject() (
...
+              postCode = "W1T 2HN"
+      )
+    )(request)
+
+  val updateUkCompanyAddress
+    : (String, JourneyRequest[AnyContent], HeaderCarrier) => Future[Address] =
+    (journeyId, request, hc) =>
+      incorpIdConnector.getDetails(journeyId)(hc)
+        .flatMap(response => updateBusinessAddress(response.companyAddress.toPptAddress)(request))
+
```
Comment (PauloPortugal) [inline app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ContactDetailsConfirmAddressController.scala:71]: I am not sure a closure provides tangible benefit here. In this case you cannot have `request` and `hr` as **implicits** and you need to explicitly call them out on the subsequent calls. 

All the benefits we get from Scala implicits are eroded by the closure. 

I can keep it the approach. Once we have more legal entities to support, potentially we can discuss the design again.

File: app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/OrganisationDetailsTypeController.scala [hunk 5]
```diff
@@ -59,21 +72,52 @@ class OrganisationDetailsTypeController @Inject() (
...
+                          case Some(OrgType.SOLE_TRADER) =>
+                            getSoleTraderRedirectUr()
+                              .map(journeyStartUrl => SeeOther(journeyStartUrl).addingToSession())
+                          case _ =>
+                            Future(
+                              Redirect(routes.OrganisationTypeNotSupportedController.onPageLoad())
+                            )
+                        }
+                      case _ => Future(Redirect(routes.RegistrationController.displayPage()))
                     }
```
Comment (PauloPortugal) [inline app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/OrganisationDetailsTypeController.scala:92]: :+1:

**Review by tmc-ee — APPROVED** [review]

File: test/base/unit/MockConnectors.scala [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    when(
+      mockSoleTraderConnector.createJourney(any[SoleTraderIncorpIdCreateRequest])(any())
+    ).thenReturn(Future.successful(redirectUrl))
+
+  def mockSoleTraderCreateIncorpJourneyIdException(): OngoingStubbing[Future[String]] =
+    when(
+      mockSoleTraderConnector.createJourney(any[SoleTraderIncorpIdCreateRequest])(any())
+    ).thenThrow(new RuntimeException("error"))
+
+  def mockukCompanyCreateIncorpJourneyId(redirectUrl: String): OngoingStubbing[Future[String]] =
```
Comment (tmc-ee) [inline test/base/unit/MockConnectors.scala:81]: Extremely petty one  but `U` should be captialised 😅

File: app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ReviewRegistrationController.scala [hunk 7]
```diff
@@ -59,16 +62,35 @@ class ReviewRegistrationController @Inject() (
...
+
+  private def ukCompanyReview(journeyId: String)(implicit request: JourneyRequest[AnyContent]) =
+    for {
+      incorporationDetails <- incorpIdConnector.getDetails(journeyId)
+      _                    <- markRegistrationAsReviewed()
+    } yield Ok(
+      page(registration = request.registration, incorporationDetails = Some(incorporationDetails))
+    )
+
   private def markRegistrationAsReviewed()(implicit
```
Comment (PauloPortugal) [inline app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ReviewRegistrationController.scala:94]: We may have to review this as well, as we are potentially going to have a lot of repetitive/duplicated code.
WE should potentially have a entityDetails (with potentially some optionals there) and in that case the code get a lot more elegant :thinking: 

We would then push difference to views :thinking:

**Review by PauloPortugal — APPROVED** [review]

File: test/base/unit/MockConnectors.scala (hunk 6 above)
Comment (tombart) [reply test/base/unit/MockConnectors.scala:81]: done

File: app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ContactDetailsConfirmAddressController.scala (hunk 2 above)
Comment (tombart) [reply app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ContactDetailsConfirmAddressController.scala:91]: ok, inlined those methods so I can use implicits

File: app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ContactDetailsConfirmAddressController.scala (hunk 3 above)
Comment (tombart) [reply app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ContactDetailsConfirmAddressController.scala:89]: ok, inlined those methods so I can use implicits

File: app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ContactDetailsConfirmAddressController.scala (hunk 4 above)
Comment (tombart) [reply app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ContactDetailsConfirmAddressController.scala:71]: yeap, I inlined those methods now and using implicits

File: app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ReviewRegistrationController.scala (hunk 7 above)
Comment (tombart) [reply app/uk/gov/hmrc/plasticpackagingtax/registration/controllers/ReviewRegistrationController.scala:94]: ok, just did enough to make sole trader work, not sure how other entities will look like yet, once know more we can bring common parts together

**Pull request merged**